Support for specifying filter preferences

danjohnson95 opened the pull request:
This PR adds functionality for a user to customise which notifications they receive.

See the full commit message for code explanations.

Users will now see a Settings cog on the top header:
<img width="405" alt="image" src="https://github.com/UserScape/larajobs-desktop/assets/6871504/b2645679-0bd2-44b4-8241-054b8de36b28">

This takes the user to the /settings route, which shows currently configured filters and allows for adding:
<img width="419" alt="image" src="https://github.com/UserScape/larajobs-desktop/assets/6871504/3147d391-cdc3-4994-aaba-2f2420e50d00">

When clicking "Add new filter", the user can specify their criteria
<img width="424" alt="image" src="https://github.com/UserScape/larajobs-desktop/assets/6871504/76245838-f4ff-49d8-80d7-9a60f73ca8e9">

<img width="419" alt="image" src="https://github.com/UserScape/larajobs-desktop/assets/6871504/ba6f58cb-8d91-46cb-ac07-cb1200f83d3e">

I've added a quick feature test for this too.

danjohnson95 committed:
feat: sticky header

danjohnson95 committed:
refact: tidy up SendNewJobsNotification listener

To simplify this listener, we're now passing a Collection of jobs into it.
This way, we can now use Collection methods such as ->count() and isEmpty().

We're using the new Notification::new() static method to instantiate
our notification client.

The name of the event when a notification is clicked now has it's
prefixed defined as a constant for consistency.

danjohnson95 committed:
feat: create resources for filtering notifications

This commit creates:

- FilterField enum, for specifiying which columns against a JobPost that can be filtered
- FilterOperation enum, for specifying how the column should be filtered
- A migration to store the user's filters
- A scoped query for applying these filters to a JobPost

danjohnson95 committed:
feat: only notify on what's been filtered

When we dispatch the JobsPosted event, the Jobs collection will only
contain job postings which meet the user's criteria.

We're applying the filtered scope on the JobPost model which applies
the user's notification preferences, or just returns the builder if
no notification preferences have been specified.

danjohnson95 committed:
chore: install Livewire

We'll use Laravel's Livewire to inject some interactivity into
the application. Initially, this will be used to allow a user
to manage their notification preferences.

danjohnson95 committed:
refact: use layout component in blade file

Previously, all the view contents were in a single view file.
I've refactored this to define a layout component, and I've
moved the header and footer into a blade component.

danjohnson95 committed:
feat: add button for Settings in header

The Header component now includes a button to take the user to the
Settings, or when the settings page is active, a Back button is
displayed instead which takes the user back to the main list.

danjohnson95 committed:
feat: add Larajobs brand colour to Tailwind colours

This now allows us to use the colour `primary` in Tailwind classnames
in order to display the Larajobs brand colour

danjohnson95 committed:
feat: create Livewire components for managing filters

This creates a Settings livewire component which lists the filters
that the user has set up, and a NewFilter modal to allow a user to
configure their filters.

danjohnson95 committed:
feat: add helper for getting friendly operation and field names

We'll use this in a component to show the user which filters
they've currently got set up.

danjohnson95 committed:
feat: create component for displaying filter info

This component displays details about each filter which the user
has added.

danjohnson95 committed:
chore: define route

This defines the route for the Settings page where a user can manage
their notification preferences

danjohnson95 committed:
refact: get jobs to notify inside event

This increases testability - we can new up an event and inspect
which jobs it's about to pass onto the listener

danjohnson95 committed:
chore: add factories

In order to test, we'll need to create some factories so we can
easily create test data.

danjohnson95 committed:
chore: set up phpunit

We'll set up PHPUnit with the sqlite database driver so we can
simulate creating jobs and filters within our test environment

danjohnson95 committed:
fix: return filters correctly

I had a couple of bugs in the switch case - it was falling through,
plus it was returning an object but the scoped query was expecting
an array of values.

danjohnson95 committed:
test: create test to assert we're notifying of relevant jobs

These tests ensure that when the JobsPosted event is constructed,
it only contains the jobs which are relevant based on the user's
filter preferences.

Unfortunately we can't mock the NativePHP Notification facade, so
instead we're just ensuring that the jobs within the event contain
what we're expecting to notify the user about.

LukeTowers commented on app/Events/JobsPosted.php:
```diff
@@ -2,18 +2,26 @@
+            ->orderBy('published_at', 'desc')
+            ->get();
     }
 }
```
```suggestion
}

```

LukeTowers commented on app/Http/Livewire/NewFilter.php:
```diff
@@ -0,0 +1,43 @@
+    {
+        return view('livewire.new-filter');
+    }
+}
```
```suggestion
}

```

LukeTowers commented on app/Http/Livewire/Settings.php:
```diff
@@ -0,0 +1,45 @@
+    {
+        $this->addNewFilter = false;
+    }
+}
```
```suggestion
}

```

LukeTowers commented on app/Jobs/FetchNewJobs.php:
```diff
@@ -61,9 +61,20 @@ public function handle(): void
+     * we'll only notify them of jobs that match their criteria.
+     *
+     * The event constructor is responsible for fetching the
+     * jobs which meet the criteria.
```
Why is the event constructor responsible for gathering the data? It seems backwards to me, you wouldn't be able to reuse the event for anything else taking that approach.

What are the benefits of making the event prepare its own data instead of passing the relevant data to the event when creating it? I'm not sure I've seen that pattern before so I'm genuinely curious.

danjohnson95 replied:
It's a good observation - I mainly did this to increase testability. I wanted our test to be able to prove that only jobs meeting the filter criteria are notified.

If we're passing the jobs to the event, we can't reliably test _which_ jobs the user will be notified of when `FetchNewJobs` is run, unless we were to copy over the logic of the query into the test (but then we'd end up with a brittle test)

I think we've got three options:

- Make the event responsible for fetching the jobs. (The change made in this PR). This way, both our tests and the real implementation will be using the same logic to fetch jobs to notify the user about. As you've mentioned, the downside of this is that the event is only usable for notifying the user of new jobs that meet their criteria and haven't already been notified of yet.
- Make the listener responsible for fetching the jobs. This would give us the same result as the above option, but at least we could re-use the event by attaching different listeners.
- Refactor the FetchNewJobs job so that it has a service responsible for fetching the data from the RSS feed. That way, we can mock the response of the RSS feed in our tests. I think that's a good change to make anyway because it'll allow us to test the persistence part of this job, but it still doesn't really give us a way to test the notification sending part _in isolation_.

What do you think? I'm leaning towards option three. We'll make the `JobsPostedEventTest` responsible for testing that notifications are sent (not _which_ jobs the user is notified of), and we'll make a new test for `FetchNewJobs` which is responsible for testing that the jobs are persisted correctly _and_ the event is dispatched with only jobs the user is interested in.

**TLDR**

- did it for testability
- thought the event should be responsible for determining if the user wants the notification.
- now I've changed my mind, I'll refactor and make a new test for testing that bit

LukeTowers replied:
The third option sounds great, thanks for thinking on it further and coming up with that as the solution :)

LukeTowers commented on app/Listeners/SendNewJobsNotification.php:
```diff
@@ -4,56 +4,78 @@
+            ->message("There are no new jobs available.")
+            ->event(self::EVENT_CLICK_PREFIX . 'nojobs')
+            ->show();
+    }
```
Good refactoring job on this class, I like it ❤️

LukeTowers commented on app/Models/Filter.php:
```diff
@@ -0,0 +1,85 @@
+    /**
+     * Transforms the filter to a query arguments
+    */
+    public function operationToQuery()
```
```suggestion
    public function operationToQuery(): array
```

LukeTowers commented on resources/views/components/filter.blade.php:
```diff
@@ -0,0 +1,5 @@
+    <span class="font-medium text-primary">{{ $filter->getFriendlyField() }}</span>
+    <span class="text-gray-500">{{ ' ' . $filter->getFriendlyOperation() . ' ' }}</span>
+    <code class="font-mono">{{ $filter->query }}</code>
+</div>
```
```suggestion
</div>

```

LukeTowers commented on resources/views/components/footer.blade.php:
```diff
@@ -0,0 +1,27 @@
+        Built by <a onclick="shell.openExternal('https://userscape.com')" href="javascript:;">UserScape</a> in partnership with <a onclick="shell.openExternal('https://laravel-news.com')" href="javascript:;">Laravel News</a>
+    </p>
+    <p class="mt-1 text-center text-gray-400">&copy; 2014 - {{ date('Y') }}</p>
+</footer>
```
```suggestion
</footer>

```

LukeTowers commented on resources/views/components/header.blade.php:
```diff
@@ -0,0 +1,35 @@
+        <a href="https://larajobs.com/laravel-consultants" class="px-4 py-2 rounded-sm border-2 border-accent bg-sky-200 text-zinc-800 hover:bg-white hover:text-accent">Hire a Consultant</a>
+        <a href="https://larajobs.com/create" class="px-4 py-2 rounded-sm border-2 border-accent bg-accent text-white hover:bg-white hover:text-accent">Post a Job</a>
+    </div> --}}
+</header>
```
```suggestion
</header>

```

LukeTowers commented on resources/views/components/layout.blade.php:
```diff
@@ -0,0 +1,25 @@
+        </div>
+        @livewireScripts
+    </body>
+</html>
```
```suggestion
</html>

```

LukeTowers commented on resources/views/livewire/new-filter.blade.php:
```diff
@@ -0,0 +1,53 @@
+                    <option value="{{ App\Enums\FilterOperation::Equals }}">equals</option>
+                    <option value="{{ App\Enums\FilterOperation::NotEquals }}">is not</option>
+                    <option value="{{ App\Enums\FilterOperation::Contains }}">contains</option>
+                    <option value="{{ App\Enums\FilterOperation::NotContains }}">doesnt contain</option>
```
```suggestion
                    <option value="{{ App\Enums\FilterOperation::NotContains }}">doesn't contain</option>
```

LukeTowers commented on resources/views/livewire/new-filter.blade.php:
```diff
@@ -0,0 +1,53 @@
+            </div>
+        </div>
+    </div>
+</div>
```
```suggestion
</div>

```

LukeTowers commented on resources/views/livewire/settings.blade.php:
```diff
@@ -0,0 +1,43 @@
+    @if ($addNewFilter)
+        <livewire:new-filter/>
+    @endif
+</div>
```
```suggestion
</div>

```

LukeTowers commented on resources/views/menubar/index.blade.php:
```diff
@@ -1,121 +1,64 @@
-</html>
+        </a>
+    @endforeach
+</x-layout>
```
```suggestion
</x-layout>

```

LukeTowers commented on routes/web.php:
```diff
@@ -19,3 +20,5 @@
     return view('menubar.index', ['jobPosts' => $jobs]);
 })->name('menubar.index');
+
+Route::get('/settings', Settings::class)->name('menubar.settings');
```
```suggestion
Route::get('/settings', Settings::class)->name('menubar.settings');

```

LukeTowers commented on tests/Feature/JobsPostedEventTest.php:
```diff
@@ -0,0 +1,55 @@
+        // notify the user about.
+        $this->assertNotContains($job->id, $event->jobs->pluck('id'));
+    }
+}
```
```suggestion
}

```

LukeTowers reviewed (changes requested):
Looks great @danjohnson95! Would you be willing to add a GitHub action to run the test suite as well? If not I should be able to take care of that

LukeTowers committed:
Apply suggestions from code review

LukeTowers committed:
Update resources/views/menubar/index.blade.php

LukeTowers committed:
Update routes/web.php

LukeTowers committed:
Update tests/Feature/JobsPostedEventTest.php

danjohnson95 committed:
fix: use Notification facade in event listener

I realised we weren't using the facade here, which is why we couldn't
mock the implementation in our tests! 🎉

danjohnson95 committed:
test: add more tests for inspecting notification contents

We're swapping out the Notification facade with a mock using
Laravel's built in facade helpers. This allows us to test
that a notification was displayed, and allows us to inspect
the contents of the notification.

danjohnson95 committed:
refact: JobsPostedEvent is now only responsible for sending notifications

Before, the event was responsible for only notifying the user
about jobs which met their criteria. But now, the event will
just notify the user of the given jobs.

It's the responsibility of the event dispatcher to only provide
jobs which the user is interested in.

danjohnson95 committed:
refact: move data fetching to separate service

Now that we're fetching the RSS feed in a separate response, the
FetchNewJobs job is no longer responsible for that. Plus, we can
now easily mock the RSS contents in tests

danjohnson95 committed:
test: write tests for FetchNewJobs class

Added tests to ensure that the JobsPostedEvent only contains
jobs that we want to notify the user of, and ensures that new
jobs are persisted in the database.

danjohnson95 committed:
Merge branch 'danjohnson95' of github.com:danjohnson95/larajobs-desktop into danjohnson95

danjohnson95 committed:
fix: add return type

Co-authored-by: Luke Towers <github@luketowers.ca>

danjohnson95 committed:
style: add missing linebreak

Co-authored-by: Luke Towers <github@luketowers.ca>

danjohnson95 committed:
feat: add workflow to run test suite on PR

danjohnson95 committed:
Merge branch 'danjohnson95' of github.com:danjohnson95/larajobs-desktop into danjohnson95

danjohnson95 commented:
Thanks for the feedback! I've added a workflow file to run the tests but it looks like it's not triggering - maybe the repo needs GitHub Actions enabled in the settings on your end? Let me know if not - it's also possible I've fudged the workflow file haha

LukeTowers commented on app/Jobs/FetchNewJobs.php:
```diff
@@ -61,8 +63,22 @@ public function handle(): void
         // Fire the event
-        // @TODO: Apply filters here
-        $jobs = JobPost::visible()->unnotified()->orderBy('published_at', 'desc')->get()->all();
+        $this->notifyJobsPosted();
```
If we're not doing anything with the $posts array then we should probably get rid of it now

danjohnson95 replied:
Good spot, thanks. It's now removed 👍

LukeTowers commented on tests/Feature/FetchNewJobsTest.php:
```diff
@@ -0,0 +1,111 @@
+            return $event->jobs->count() === 0;
+        });
+    }
+}
```
```suggestion
}

```

LukeTowers committed:
Update tests/Feature/FetchNewJobsTest.php

LukeTowers committed:
Remove unused variable
